sklearn/tree/_tree.pyx

```diff
@@ -179,9 +181,14 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
         cdef bint is_left
         cdef SIZE_t n_node_samples = splitter.n_samples
         cdef double weighted_n_node_samples
-        cdef SplitRecord split
         cdef SIZE_t node_id
 
+        # initialize record to keep track of split node data and a pointer to the
+        # memory address containing the split node
+        # Note: the pointer allows us to modularly define different split records
+        cdef SplitRecord split
+        cdef SplitRecord* split_ptr = <SplitRecord *>malloc(splitter.pointer_size())
```

Comment:
Allocating and freeing memory repeatedly and manually is super costly and prone to memoryleaks. I think we would rather be preferable to have `Splitter` return a value of a Cython extension type which extend `SplitRecord`, pass it by address and have method in other `Tree` cast record to to the proper extension type extending `SplitRecord`.

What do you think, @adam2392?

Reply:
Hi @jjerphan do you mean SplitRecord would be refactored to a cdef class and Splitter returns a pointer/address to the class inside a function?

E.g.

```
cdef class SplitRecord:
    # Data to track sample split
    cdef SIZE_t feature         # Which feature to split on.
    cdef SIZE_t pos             # Split samples array at the given position,
                           # i.e. count of samples below threshold for feature.
                           # pos is >= end if the node is a leaf.
    cdef double threshold       # Threshold to split at.
    cdef double improvement     # Impurity improvement given parent node.
    cdef double impurity_left   # Impurity of the left split.
    cdef double impurity_right  # Impurity of the right split.

cdef class Splitter:
    ...
    cdef SplitRecord split_record_type(self):
            return SplitRecord

cdef class Tree:
    cdef void* get_split_record(self):
           # cast splitRecord pointer to the proper extension type
```

Is it okay to replace `SplitRecord` struct with a class? If so, it would make our life way easier I think.

Reply:
Hi, we probably will need to have something like this, yes.

I need to scope some proper time to look at the current design and implementations to get a sense of what would be the most adapted.

Reply:
I don't think that is possible because `cdef class SplitRecord` is considered a "Python object" and cannot be instantiated within a `cdef func(...) nogil` function. 

If we are concerned w/ memory allocation, then perhaps instead of pure malloc, we can use the C-API, which is exposed inside Cython: https://cython.readthedocs.io/en/latest/src/tutorial/memory_allocation.html

E.g.

```
# Tree code that stays pretty much the same except adding this pointer logic specifically, so
# the code allows modularity wrt `SplitRecord`
def __cinit__(self):
        # allocate some memory (uninitialised, may contain arbitrary data)
        self.split_record = <double*> PyMem_Malloc(
            self.pointer_size())
        if not self.split_record:
            raise MemoryError()
    def __dealloc__(self):
        PyMem_Free(self.split_record)  # no-op if self.data is NULL

    cdef _add_split_node(...) nogil:
          cdef SplitRecord split
          cdef SplitRecord* split_ptr = <SplitRecord *>malloc(splitter.pointer_size())

          splitter.node_split(..., split_ptr)

...
      # some subclass that wants to subclass the SplitRecord
      cdef node_split(..., SplitRecord* split_ptr):
            # typecast the pointer to an ObliqueSplitRecord
            cdef ObliqueSplitRecord* oblique_split = <ObliqueSplitRecord*>(split)
```

Reply:
@jjerphan  

Unfortunately, Cython extension types as a class cannot be "instantiated" within a nogil block since they are Python objects. However, a possible solution is to instantiate the `SplitRecord` for `best` and `current` inside Splitter's `__cinit__` function. Then they can be "used" within the nogil blocks. The other areas that `SplitRecord` is used are not within nogil functions, so they can be safely instantiated.

This would answer the question of how we replace the struct with an extension type. The next challenge is how do we support modularity?

If they are instantiated within core functions like `TreeBuilder.build()`, then one would need to overwrite the entire build function if they use a new Splitter/SplitRecord. This is rather undesirable and defeats the purpose of replacing the struct w/ the extension type in the first place.

Reference: https://stackoverflow.com/questions/74749054/how-when-to-use-a-cython-extension-type-vs-a-cython-struct-to-store-data-that-is/74753005?noredirect=1#comment131993003_74753005

Reply:
@jjerphan actually, https://github.com/scikit-learn/scikit-learn/blob/a7cd0ca44d0af64bc22a7213371e3c11d94dbbe8/sklearn/ensemble/_hist_gradient_boosting/splitting.pyx#L487-L488 has this feature inside of it and it seems fine? I think there is no other way around it from my research into this problem and it is the recommended way of "subclassing" a struct. WDYT?